openedx/features/offline_content/docs/001-mobile-offline-content-support.rst

```diff
@@ -0,0 +1,134 @@
+Every time block content republished ZIP archive with offline content should be regenerated.
+The xBlock should be rendered the same way it’s rendered for a student using /xblock/{locator.id} endpoint in LMS.
+HTML should be processed, all related assets files, images and scripts should be included in the generated ZIP archive with offline content
+The Generation process should work with local media storage as well as s3.
+If error retrieving block happened, the generation task will be scheduled for retry 2 more times, with progressive delay.
+
+    .. image:: _images/mobile_offline_content_generation.svg
+        :alt: Mobile Offline Content Generation Process Diagram
+
+Mobile API extension
+~~~~~~~~~~~~~~~~~~~~
+
+Extend mobile API endpoint for Course Home, to return information about offline content available for download for supported blocks
+{
+...
+"offline_download": {
+    "file_url": "{file_url}" or null,
+    "last_modified": "{DT}" or null,
+    "file_size": ""
+  }
+}
+
+JavaScript Bridge for interaction with mobile applications
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Implement JS Bridge JS script to intercept and send results to mobile device for supported CAPA problems
+The submission data should be sent via bridge to IOS and Android devices.
+This script should expose markCompleted JS function so mobile can change state of the offline problem after the data was saved into internal database or on initialization of the problem
+
+
+* **Implement of a mechanism for generating and storing on a server or external storage**: The course content should be pre-generated and saved to the storage for later download.
+    * **Render content**: Generate HTML content of block as it does for LMS.
+    * **Replace static and media**: Save static and media assets files used in block to temporary directory and replace their static paths with local paths.
+    * **Archive and store content**: Archive the generated content and store it on the server or external storage.
+* **Mechanism for updating the generated data**: When updating course blocks (namely when publishing) the content that has been changed should be re-generated.
+    * **Track course publishing events on CMS side**: Signal in the CMS that makes request to LMS to update course content.
+    * **Track course publishing events on LMS side**: API endpoint to receive the signal from CMS and update course content.
+    * **Update archive**: Check generated archive creation date and update it if less than course publishing date.
+* **Implement a Mobile Local Storage Mechanism**: Use the device's local storage to save course content for offline access.
+    * **Extend blocks API**: Add links to download blocks content and  where it is possible.
+* **Sync Mechanism**: Periodically synchronize local data with the server when the device is online.
+    * **Sync on app side**: On course outline screen, check if the course content is up to date and update it if necessary.
```

Comment:
Please say more about:
* How the offline app will respond when a user hits Submit
* What local state the app will store and in what format it will store it
* How the app will synchronize this state with the server

Reply:
The JS bridge will intercept ajax requests in the mobile application and store the responses. The user will see the message "Your response is accepted" and the Submit button should be disabled.
Then, when the user gets the Internet, the mobile application will send the saved responses to the backend one by one, to the standard xblock handler. 
So far, this is the behavior planned within FC-0047, which should be improved in the future within FC-7879, where it is planned to add the ability for the application to send all responses to the backend at once.

Reply:
> The JS bridge will intercept ajax requests in the mobile application and store the responses. The user will see the message "Your response is accepted" and the Submit button should be disabled.
Then, when the user gets the Internet, the mobile application will send the saved responses to the backend one by one, to the standard xblock handler.

Can you add these ^ details to the ADR?